repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.atlas.type.AtlasTypeRegistry;
+import org.apache.atlas.utils.AtlasPerfMetrics;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.collections.MapUtils;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+import static org.apache.atlas.repository.Constants.*;
+import static org.apache.atlas.repository.store.graph.v2.preprocessor.PreProcessorUtils.*;
+import static org.apache.atlas.repository.util.AccessControlUtils.ATTR_POLICY_RESOURCES;
+import static org.apache.atlas.repository.util.AtlasEntityUtils.mapOf;
+
+public class DataProductPreProcessor extends AbstractDomainPreProcessor {
+    private static final Logger LOG = LoggerFactory.getLogger(DomainPreProcessor.class);
+    private AtlasEntityHeader parentDomain;
+    private EntityMutationContext context;
+    public DataProductPreProcessor(AtlasTypeRegistry typeRegistry, EntityGraphRetriever entityRetriever,
+                                   AtlasGraph graph, EntityGraphMapper entityGraphMapper) {
+        super(typeRegistry, entityRetriever, graph);
+    }
+
+    @Override
+    public void processAttributes(AtlasStruct entityStruct, EntityMutationContext context,
+                                  EntityMutations.EntityOperation operation) throws AtlasBaseException {
+        //Handle name & qualifiedName
+        if (operation == EntityMutations.EntityOperation.CREATE && LOG.isDebugEnabled()) {
+            LOG.debug("DataProductPreProcessor.processAttributes: pre processing {}, {}",
+                    entityStruct.getAttribute(QUALIFIED_NAME), operation);
+        }
+        this.context = context;
+
+        AtlasEntity entity = (AtlasEntity) entityStruct;
+        AtlasVertex vertex = context.getVertex(entity.getGuid());
+
+        setParent(entity, context);
+
+        if (operation == EntityMutations.EntityOperation.CREATE) {
+            processCreateProduct(entity, vertex);
+        }
+    }
+
+    private void processCreateProduct(AtlasEntity entity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processCreateDomain");
+        String domainName = (String) entity.getAttribute(NAME);
+        String parentDomainQualifiedName = (String) entity.getAttribute(PARENT_DOMAIN_QN);
+
+        productExists(domainName, parentDomainQualifiedName);
+        String newQualifiedName = createQualifiedName(parentDomainQualifiedName);
+        if(!newQualifiedName.isEmpty()){
+            entity.setAttribute(QUALIFIED_NAME, newQualifiedName);
+        }
+        else{
+            throw new AtlasBaseException(AtlasErrorCode.BAD_REQUEST, "Parent Domain Qualified Name is empty");
+        }
+
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    public static String createQualifiedName(String parentDomainQualifiedName) {
```

Comment:
We should always expect parentDomainQualifiedName here, if empty or null throw an exception

Reply:
No need to check parentDomainQualifiedName != null 
StringUtils.isNotEmpty(parentDomainQualifiedName) takes care of both null & empty